### PR TITLE
fwupd: Fix daemon config not being found

### PR DESCRIPTION
**Pull request by infinisil**
###### Motivation for this change
Fixes https://github.com/NixOS/nixpkgs/issues/48425
Fix by @dtzWill mentioned in https://github.com/NixOS/nixpkgs/pull/47057

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/os-specific/linux/firmware/fwupd/default.nix [hunk 1]
```diff
@@ -46,6 +46,11 @@ in stdenv.mkDerivation {
 
     patchShebangs .
     substituteInPlace data/installed-tests/fwupdmgr.test.in --subst-var-by installedtestsdir "$installedTests/share/installed-tests/fwupd"
+
+    # /etc/daemon.conf
+    substituteInPlace meson.build --replace \
```
Comment (jtojnar) [inline pkgs/os-specific/linux/firmware/fwupd/default.nix]: Could you use a patch file instead? It will be more clear when the pattern no longer matches.

Comment (Infinisil) [reply pkgs/os-specific/linux/firmware/fwupd/default.nix]: I also just realized that there's a patch touching the config file already: https://github.com/NixOS/nixpkgs/blob/4207f01f27fb6f3ba22c4720d529812601ad87f1/pkgs/os-specific/linux/firmware/fwupd/fix-paths.patch#L12-L17

I'll investigate further later

**Comment by jtojnar** [comment]
Something like this ought to do it:

```patch
diff --git a/pkgs/os-specific/linux/firmware/fwupd/fix-paths.patch b/pkgs/os-specific/linux/firmware/fwupd/fix-paths.patch
index 1cd9eb0e8f1..212f30147c2 100644
--- a/pkgs/os-specific/linux/firmware/fwupd/fix-paths.patch
+++ b/pkgs/os-specific/linux/firmware/fwupd/fix-paths.patch
@@ -1,89 +1,49 @@
+diff --git a/data/builder/meson.build b/data/builder/meson.build
+index c7a430c0..e69de29b 100644
 --- a/data/builder/meson.build
 +++ b/data/builder/meson.build
 @@ -1,3 +0,0 @@
 -install_data('README.md',
 -  install_dir : join_paths(localstatedir, 'lib', 'fwupd', 'builder')
 -)
---- a/data/meson.build
-+++ b/data/meson.build
-@@ -8,16 +8,12 @@
-   subdir('installed-tests')
- endif
+diff --git a/meson.build b/meson.build
+index 96d752ae..fc5cb4dd 100644
+--- a/meson.build
++++ b/meson.build
+@@ -294,11 +294,17 @@
  
--install_data(['daemon.conf'],
--  install_dir : join_paths(sysconfdir, 'fwupd')
--)
--
- install_data(['org.freedesktop.fwupd.metainfo.xml'],
-   install_dir: join_paths(datadir, 'metainfo')
- )
- 
- install_data(['org.freedesktop.fwupd.conf'],
--  install_dir : join_paths(sysconfdir, 'dbus-1', 'system.d')
-+  install_dir : join_paths(get_option('prefix'), 'etc', 'dbus-1', 'system.d')
- )
+ plugin_dir = join_paths(libdir, 'fwupd-plugins-3')
  
- install_data(['metadata.xml'],
---- a/data/pki/meson.build
-+++ b/data/pki/meson.build
-@@ -4,14 +4,14 @@
-       'GPG-KEY-Linux-Foundation-Firmware',
-       'GPG-KEY-Linux-Vendor-Firmware-Service',
-     ],
--    install_dir : join_paths(sysconfdir, 'pki', 'fwupd')
-+    install_dir : join_paths(get_option('prefix'), 'etc', 'pki', 'fwupd')
-   )
++default_sysconfdir = get_option('sysconfdir')
++if default_sysconfdir == 'etc'
++  message('sysconfdir of etc makes no sense, using /etc')
++  default_sysconfdir = '/etc'
++endif
++
+ conf.set_quoted('BINDIR', bindir)
+ conf.set_quoted('LIBEXECDIR', libexecdir)
+ conf.set_quoted('DATADIR', datadir)
+ conf.set_quoted('LOCALSTATEDIR', localstatedir)
+-conf.set_quoted('SYSCONFDIR', sysconfdir)
++conf.set_quoted('SYSCONFDIR', default_sysconfdir)
+ conf.set_quoted('PLUGINDIR', plugin_dir)
  
-   install_data([
-       'GPG-KEY-Linux-Foundation-Metadata',
-       'GPG-KEY-Linux-Vendor-Firmware-Service',
-     ],
--    install_dir : join_paths(sysconfdir, 'pki', 'fwupd-metadata')
-+    install_dir : join_paths(get_option('prefix'), 'etc', 'pki', 'fwupd-metadata')
-   )
- endif
- 
-@@ -19,12 +19,12 @@
-   install_data([
-       'LVFS-CA.pem',
-     ],
--    install_dir : join_paths(sysconfdir, 'pki', 'fwupd')
-+    install_dir : join_paths(get_option('prefix'), 'etc', 'pki', 'fwupd')
-   )
-   install_data([
-       'LVFS-CA.pem',
-     ],
--    install_dir : join_paths(sysconfdir, 'pki', 'fwupd-metadata')
-+    install_dir : join_paths(get_option('prefix'), 'etc', 'pki', 'fwupd-metadata')
-   )
- endif
- 
---- a/data/remotes.d/meson.build
-+++ b/data/remotes.d/meson.build
-@@ -3,7 +3,7 @@
-       'lvfs.conf',
-       'lvfs-testing.conf',
-     ],
--    install_dir : join_paths(sysconfdir, 'fwupd', 'remotes.d')
-+    install_dir : join_paths(get_option('prefix'), 'etc', 'fwupd', 'remotes.d')
-   )
-   i18n.merge_file(
-     input: 'lvfs.metainfo.xml',
-@@ -37,12 +37,12 @@
-   output : 'fwupd.conf',
-   configuration : con2,
-   install: true,
--  install_dir: join_paths(sysconfdir, 'fwupd', 'remotes.d'),
-+  install_dir: join_paths(get_option('prefix'), 'etc', 'fwupd', 'remotes.d'),
- )
- configure_file(
-   input : 'vendor.conf',
-   output : 'vendor.conf',
-   configuration : con2,
-   install: true,
--  install_dir: join_paths(sysconfdir, 'fwupd', 'remotes.d'),
-+  install_dir: join_paths(get_option('prefix'), 'etc', 'fwupd', 'remotes.d'),
+ conf.set_quoted('GETTEXT_PACKAGE', meson.project_name())
+@@ -310,12 +316,6 @@
+   configuration : conf
  )
+ 
+-default_sysconfdir = get_option('sysconfdir')
+-if default_sysconfdir == 'etc'
+-  message('sysconfdir of etc makes no sense, using /etc')
+-  default_sysconfdir = '/etc'
+-endif
+-
+ plugin_deps = []
+ plugin_deps += appstream_glib
+ plugin_deps += gio
+diff --git a/meson_post_install.sh b/meson_post_install.sh
+index 0cbb6f41..d757a81a 100755
 --- a/meson_post_install.sh
 +++ b/meson_post_install.sh
 @@ -11,6 +11,4 @@

```

**Comment by Infinisil** [comment]
@jtojnar Um, could you please send the file directly instead of a diff of it (or commit it directly)? I can't apply it because git apply/patch isn't happy with it being a diff of a diff :P

**Comment by jtojnar** [comment]
I do not see how would that be a problem, only the first character on each line should be considered. Try applying it to unstable.

**Comment by Infinisil** [comment]
@jtojnar Ah yeah it worked, I just needed `patch pkgs/os-specific/firmware/fwupd/fix-paths.patch the.patch`

I'll test this PR again this evening (by just running the fwupd service).

**Comment by jtojnar** [comment]
@dtzWill at https://github.com/NixOS/nixpkgs/pull/45950#issuecomment-428405545 wrote:

> NixOS's extraUtils fails with this due to reference that's not allowed-- at first I feared this was the same problem as before (since the symptom is the same as when we were corrupting things re:meson), but I think it's unrelated :). Regardless I suspect you're running into this and if you haven't you probably will (or have fixed it and I'm on wrong branch :D).
> 
> In short the fix is adding `-Dlink-udev-shared=false` to systemd's mesonFlags. I didn't chase down why it suddenly started happening, but I can say I noticed in past few months various other distributions making changes regarding this so perhaps it's due to a meson upgrade they did before us :).
> 
> The harder fix, and untested, is that the extraUtils find-libs-and-copy-and-patchelf stuff is both too complicated and in particular fails to account for transitive dependencies (libraries that need other libraries). Well "fails" may be incorrect, this is possibly a feature and some would argue a case of underlinking but all I mean is this wouldn't happen if the xz dep was pulled in and patched to be used.
> 
> I think the solution proposed (meson flags) is the best as it prevents general systemd deps (== take over the world) from being pulled into our early boot process. We may want to do the same for systemctl (similar option) but that's not necessary so I'll leave that for another time.
> 
> Anyway thanks for your work, hope this helps, LMK if you have any questions or whatnot :).

I still cannot reproduce this failure, when trying to build the following expression:

```nix
{pkgs ? import <nixpkgs> {} }:
let
  config.boot.loader.supportsInitrdSecrets = false;
  config.boot.isContainer = false;
  config.boot.initrd.secrets = {};
  config.boot.initrd.extraUtilsCommands = "";
  config.boot.initrd.extraUtilsCommandsTest = "";
  config.system.build.fileSystems = [];
  config.systemd.package = pkgs.systemd;
  inherit (pkgs) lib;

  utils = import <nixpkgs/nixos/lib/utils.nix> pkgs;
  stage-1 = import <nixpkgs/nixos/modules/system/boot/stage-1.nix> { inherit pkgs lib utils config; };

in stage-1.config.content.system.build.extraUtils
```

**Comment by jtojnar** [comment]
Could you add a comment with link to next to the patch? https://github.com/hughsie/fwupd/issues/800

File: pkgs/os-specific/linux/firmware/fwupd/default.nix [hunk 2]
```diff
@@ -36,6 +36,7 @@ in stdenv.mkDerivation {
   LC_ALL = "en_US.UTF-8"; # For po/make-images
 
   patches = [
+    # https://github.com/hughsie/fwupd/issues/800
```
Comment (Infinisil) [inline pkgs/os-specific/linux/firmware/fwupd/default.nix]: @jtojnar Like this?

**Comment by jtojnar** [comment]
New patch: https://github.com/hughsie/fwupd/pull/824

**Comment by jtojnar** [comment]
Closing in favour of https://github.com/NixOS/nixpkgs/pull/48519

**Pull request closed without merging**